llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -7687,6 +7688,29 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
   if (Instruction *Res = foldReductionIdiom(I, Builder, DL))
     return Res;
 
+  {
+    Value *A, *B;
+    const APInt *C;
+    ICmpInst::Predicate PredEq = ICmpInst::ICMP_EQ;
+    if (I.getPredicate() == PredEq) {
+      if (match(Op0, m_And(m_Value(A), m_APInt(C))) && match(Op1, m_One())) {
```

Comment:
Just matching `m_And(m_SExtLike(m_Value(A)...)` seems simpler.

Note `m_SExtLike` just matches `sext` or a `zext` that is provably equivilent to a `sext` due to the `nneg` IR flag.